File: ide-common/src/main/java/org/digma/intellij/plugin/document/CodeLensProvider.java
package org.digma.intellij.plugin.document;

import com.intellij.openapi.Disposable;
import com.intellij.openapi.diagnostic.Logger;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.vfs.VirtualFile;
import org.digma.intellij.plugin.analytics.*;
import org.digma.intellij.plugin.codelens.CodeLensProviderRefresh;
import org.digma.intellij.plugin.common.*;
import org.digma.intellij.plugin.errorreporting.ErrorReporter;
import org.digma.intellij.plugin.log.Log;
import org.digma.intellij.plugin.model.discovery.*;
import org.digma.intellij.plugin.model.lens.CodeLens;
import org.digma.intellij.plugin.model.rest.codelens.*;
import org.digma.intellij.plugin.model.rest.insights.*;
import org.jetbrains.annotations.NotNull;

import java.util.*;
import java.util.concurrent.ConcurrentHashMap;


/**
 * get code lens from backend , builds it and caches it while the file is opened.
 * listens to documentInfoChanged and builds the code lens on background.
 * documentInfoChanged will be fired when a file is opened and every time it is changed.
 * when the file is closed the code lens are removed.
 * consumers for code lens never wait for the backend and always get what CodeLensProvider has in its cache.
 * the cache is refreshed every some seconds by CodeLensProviderRefresh that calls refresh on background.
 */
public class CodeLensProvider implements Disposable {

    private static final Logger LOGGER = Logger.getInstance(CodeLensProvider.class);

    private final Project project;

    private final Map<VirtualFile, DocumentLensPair> codeLensPerFile = new ConcurrentHashMap<>();


    public CodeLensProvider(Project project) {
        this.project = project;
        //start CodeLensProviderRefresh. CodeLensProviderRefresh is in kotlin so can use coroutines
        new CodeLensProviderRefresh(project, this).start();
    }

    @Override
    public void dispose() {
        codeLensPerFile.clear();
    }

    public static CodeLensProvider getInstance(Project project) {
        return project.getService(CodeLensProvider.class);
    }


    @NotNull
    public Set<CodeLens> provideCodeLens(@NotNull VirtualFile file) {
        Log.log(LOGGER::trace, "Got request for code lens for {}", file);
        var lensPerFile = codeLensPerFile.get(file);
        if (lensPerFile == null) {
            Log.log(LOGGER::trace, "No lenses for file in cache, returning empty {}", file);
            return Collections.emptySet();
        }
        Log.log(LOGGER::trace, "returning lenses for file {} [{}]", file, lensPerFile.codeLensList);
        return lensPerFile.codeLensList;
    }

    public void clearCodeLens(@NotNull VirtualFile file) {
        Log.log(LOGGER::trace, "clearing code lens for {}", file);
        codeLensPerFile.remove(file);
    }


    /**
     * called on background when a file is opened or changed and build code lest for the file.
     * code lens are kept in local cache.
     * returns true is code lens changed.
     */
    public boolean buildCodeLens(@NotNull VirtualFile file, @NotNull DocumentInfo documentInfo) throws AnalyticsServiceException {
        var oldLenses = provideCodeLens(file);
        buildCodeLensImpl(file, documentInfo);
        var newLenses = provideCodeLens(file);
        return !oldLenses.equals(newLenses);
    }


    private void buildCodeLensImpl(@NotNull VirtualFile file, @NotNull DocumentInfo documentInfo) throws AnalyticsServiceException {

        Log.log(LOGGER::trace, "Building code lens for {}", file);

        if (!VfsUtilsKt.isValidVirtualFile(file) || !ProjectUtilsKt.isProjectValid(project)) {
            return;
        }

        EDT.assertNonDispatchThread();
        ReadActions.assertNotInReadAccess();

        var codeLens = buildCodeLens(documentInfo);
        Log.log(LOGGER::trace, "Built code lens for {}, [{}]", file, codeLens);
        codeLensPerFile.put(file, new DocumentLensPair(documentInfo, codeLens));
    }


    /**
     * called on background by refresh task and when environment changed.
     * refresh code lens for the documents currently in the cache.
     * return the list of psi urls where code lens changed
     */
    public List<VirtualFile> refresh() {

        Log.log(LOGGER::trace, "Got request to refresh code lens");

        EDT.assertNonDispatchThread();
        ReadActions.assertNotInReadAccess();

        var changed = new ArrayList<VirtualFile>();

        codeLensPerFile.forEach((file, documentLensPair) -> {
            try {
                var documentInfo = documentLensPair.documentInfo;
                Log.log(LOGGER::trace, "refreshing code lens for {}", file);
                var oldLenses = documentLensPair.codeLensList;
                var newLenses = buildCodeLens(documentInfo);
                documentLensPair.codeLensList = newLenses;
                if (!oldLenses.equals(newLenses)) {
                    Log.log(LOGGER::trace, "Got refreshed code lens for {}, {}", file, newLenses);
                    changed.add(file);
                }

            } catch (Throwable e) {
                Log.warnWithException(LOGGER, project, e, "error in code lens refresh");
                ErrorReporter.getInstance().reportError(project, "CodeLensProvider.refresh", e);
            }
        });

        return changed;
    }


    //synchronized to prevent multiple threads building code lens. it all happens on background
    // and should not impact performance of the plugin. consumers of code lens never wait and take what is already
    // in the codeLensPerFile cache.
    @NotNull
    private synchronized Set<CodeLens> buildCodeLens(@NotNull DocumentInfo documentInfo) throws AnalyticsServiceException {

        //LinkedHashSet retains insertion order so retains the order from backend
        Set<CodeLens> codeLensList = new LinkedHashSet<>();

        var methodsInfo = documentInfo.getMethods().values();

        List<MethodWithCodeObjects> methods = new ArrayList<>();

        for (MethodInfo methodInfo : methodsInfo) {
            List<String> relatedSpansCodeObjectIds = methodInfo.getSpans().stream().map(SpanInfo::idWithType).toList();
            List<String> relatedEndpointCodeObjectIds = methodInfo.getEndpoints().stream().map(EndpointInfo::getId).toList();

            for (String id : methodInfo.allIdsWithType()) {
                methods.add(new MethodWithCodeObjects(id, relatedSpansCodeObjectIds, relatedEndpointCodeObjectIds));
            }
        }

        var methodsWithCodeLens = AnalyticsService.getInstance(project).getCodeLensByMethods(methods).getMethodWithCodeLens();

        for (MethodWithCodeLens methodWithCodeLens : methodsWithCodeLens) {
            var codeObjectId = CodeObjectsUtil.stripMethodPrefix(methodWithCodeLens.getMethodCodeObjectId());
            var decorators = methodWithCodeLens.getDecorators();

            var liveDecorator =
                    decorators.stream().filter(d -> d.getTitle().equals("Live")).findFirst()
                            .orElse(null);

            if (liveDecorator != null) {
                var codeLens = buildCodeLensOfActive(codeObjectId, liveDecorator);
                decorators.remove(liveDecorator);
                codeLensList.add(codeLens);
            }

            for (Decorator decorator : decorators) {

                int importance = decorator.getImportance().getPriority();

                String priorityEmoji = "";
                if (isImportant(importance)) {
                    priorityEmoji = "❗️";
                }

                String title = priorityEmoji + decorator.getTitle();

                //title is used as id of CodeLens
                CodeLens codeLens = new CodeLens(decorator.getTitle(), codeObjectId, decorator.getScopeCodeObjectId(), title, importance);
                codeLens.setLensDescription(decorator.getDescription());
                codeLens.setLensMoreText("Go to " + title);

                codeLensList.add(codeLens);
            }

        }

        return codeLensList;
    }

    private static CodeLens buildCodeLensOfActive(String methodId, Decorator liveDecorator) {
        var title = Unicodes.getLIVE_CIRCLE();
        CodeLens codeLens = new CodeLens(liveDecorator.getTitle(), methodId, liveDecorator.getScopeCodeObjectId(), title, 1);
        codeLens.setLensDescription(liveDecorator.getDescription());

        return codeLens;
    }

    private static boolean isImportant(Integer importanceLevel) {
        return importanceLevel <= InsightImportance.HighlyImportant.getPriority() && importanceLevel >= InsightImportance.ShowStopper.getPriority();
    }


    private static class DocumentLensPair {
        private final @NotNull DocumentInfo documentInfo;
        private @NotNull Set<CodeLens> codeLensList;

        public DocumentLensPair(@NotNull DocumentInfo documentInfo, @NotNull Set<CodeLens> codeLensList) {
            this.documentInfo = documentInfo;
            this.codeLensList = codeLensList;
        }
    }

}

File: ide-common/src/main/kotlin/org/digma/intellij/plugin/codelens/provider/CodeLensProvider.kt
package org.digma.intellij.plugin.codelens.provider

import com.intellij.openapi.Disposable
import com.intellij.openapi.components.Service
import com.intellij.openapi.components.service
import com.intellij.openapi.diagnostic.Logger
import com.intellij.openapi.project.Project
import com.intellij.openapi.vfs.VirtualFile
import io.ktor.util.collections.ConcurrentMap
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.ensureActive
import org.digma.intellij.plugin.analytics.BackendConnectionMonitor
import org.digma.intellij.plugin.common.isValidVirtualFile
import org.digma.intellij.plugin.document.DocumentInfoStorage
import org.digma.intellij.plugin.kotlin.ext.launchWhileActiveWithErrorReporting
import org.digma.intellij.plugin.kotlin.ext.launchWithErrorReporting
import org.digma.intellij.plugin.log.Log
import org.digma.intellij.plugin.model.discovery.DocumentInfo
import org.digma.intellij.plugin.model.lens.CodeLens
import java.util.concurrent.ConcurrentHashMap
import kotlin.coroutines.coroutineContext
import kotlin.time.Duration.Companion.minutes

@Service(Service.Level.PROJECT)
class CodeLensProvider(private val project: Project, private val cs: CoroutineScope) : Disposable {

    private val logger = Logger.getInstance(this::class.java)

    private val codeLensBuilder = CodeLensBuilder(project)
    private val codeLensCache: ConcurrentMap<VirtualFile, Set<CodeLens>> = ConcurrentMap()
    private val runningLoadJobs: ConcurrentMap<VirtualFile, Job> = ConcurrentMap()
    private val runningRefreshJobs: MutableSet<Job> = ConcurrentHashMap.newKeySet()

    companion object {
        @JvmStatic
        fun getInstance(project: Project): CodeLensProvider {
            return project.service<CodeLensProvider>()
        }
    }

    init {
        cs.launchWhileActiveWithErrorReporting(1.minutes,1.minutes,true,"CodeLensProvider.refreshLoop", logger) {
            refresh()
        }
    }

    override fun dispose() {
        runningLoadJobs.values.forEach { it.cancel(CancellationException("CodeLensProvider is disposing")) }
        runningRefreshJobs.forEach { it.cancel(CancellationException("CodeLensProvider is disposing")) }
        codeLensCache.clear()
    }

    internal fun loadCodeLens(file: VirtualFile, documentInfo: DocumentInfo) {
        //don't launch if there is no connection to backend
        if (BackendConnectionMonitor.getInstance(project).isConnectionError()) {
            if (logger.isTraceEnabled) {
                Log.log(logger::trace, "loading code lens for file called but no connection {}", file)
            }
            return
        }

        if (logger.isTraceEnabled) {
            Log.log(logger::trace, "starting loadCodeLens job for {}", file)
        }
        runningLoadJobs[file]?.cancel(CancellationException("CodeLensProvider.loadCodeLens called again before previous job finished"))
        val job = cs.launchWithErrorReporting("CodeLensProvider.loadCodeLens", logger) {
            if (!isValidVirtualFile(file)) return@launchWithErrorReporting
            loadCodeLensInternal(file, documentInfo)
        }
        runningLoadJobs[file] = job
        job.invokeOnCompletion { cause ->
            runningLoadJobs.remove(file)
        }
    }

    private suspend fun loadCodeLensInternal(file: VirtualFile, documentInfo: DocumentInfo) {
        coroutineContext.ensureActive()
        if (logger.isTraceEnabled) {
            Log.log(logger::trace, "loading code lens for file {}", file)
        }
        val codeLens = codeLensBuilder.buildCodeLens(documentInfo)
        if (logger.isTraceEnabled) {
            Log.log(logger::trace, "code lens for file {} is {}", file, codeLens)
        }
        coroutineContext.ensureActive()
        val previousCodeLens = codeLensCache[file]
        codeLensCache[file] = codeLens
        if (previousCodeLens != codeLens) {
            coroutineContext.ensureActive()
            if (logger.isTraceEnabled) {
                Log.log(logger::trace, "code lens changed for file {}", file)
            }
            project.messageBus.syncPublisher(CodeLensChanged.CODELENS_CHANGED_TOPIC).codelensChanged(file)
        }
    }


    internal fun removeCodeLens(file: VirtualFile) {
        if (logger.isTraceEnabled) {
            Log.log(logger::trace, "starting removeCodeLens job for {}", file)
        }
        runningLoadJobs[file]?.cancel()
        cs.launchWithErrorReporting("CodeLensProvider.removeCodeLens", logger) {
            if (logger.isTraceEnabled) {
                Log.log(logger::trace, "removing code lens for file {}", file)
            }
            codeLensCache.remove(file)
            project.messageBus.syncPublisher(CodeLensChanged.CODELENS_CHANGED_TOPIC).codelensRemoved(file)
        }
    }

    internal fun clearCodeLens() {
        if (logger.isTraceEnabled) {
            Log.log(logger::trace, "starting clearCodeLens job")
        }
        cs.launchWithErrorReporting("CodeLensProvider.clearCodeLens", logger) {
            Log.log(logger::trace, "clearing code lens")
            runningLoadJobs.values.forEach { it.cancel(CancellationException("CodeLensProvider.clearCodeLens called before previous job finished")) }
            runningRefreshJobs.forEach { it.cancel(CancellationException("CodeLensProvider.clearCodeLens called before previous job finished")) }
            codeLensCache.clear()
            project.messageBus.syncPublisher(CodeLensChanged.CODELENS_CHANGED_TOPIC).codelensCleared()
        }
    }

    internal fun refresh() {
        //don't launch if there is no connection to backend
        if (BackendConnectionMonitor.getInstance(project).isConnectionError()) {
            if (logger.isTraceEnabled) {
                Log.log(logger::trace, "refresh code lens for called but no connection")
            }
            return
        }

        runningRefreshJobs.forEach { it.cancel(CancellationException("CodeLensProvider.refresh called again before previous job finished")) }
        val job = cs.launchWithErrorReporting("CodeLensProvider.refresh", logger) {
            if (logger.isTraceEnabled) {
                Log.log(logger::trace, "refreshing code lens")
            }
            val files = codeLensCache.keys.toList()
            files.forEach { file ->
                val documentInfo = DocumentInfoStorage.getInstance(project).getDocumentInfo(file)
                if (documentInfo != null) {
                    if (logger.isTraceEnabled) {
                        Log.log(logger::trace, "refreshing code lens for file {}", file)
                    }
                    loadCodeLens(file, documentInfo)
                } else {
                    if (logger.isTraceEnabled) {
                        Log.log(logger::trace, "no document info to refresh for file {}. removing code lens", file)
                    }
                    removeCodeLens(file)
                }
            }
        }
        runningRefreshJobs.add(job)
        job.invokeOnCompletion { cause ->
            runningRefreshJobs.remove(job)
        }
    }

    fun getCodeLens(file: VirtualFile): Set<CodeLens> {
        return codeLensCache.getOrPut(file) { emptySet() }
    }
}